src/FSharp.Data.GraphQL.Server/Schema.fs

```diff
@@ -177,6 +177,41 @@ type Schema<'Root> (query: ObjectDef<'Root>, ?mutation: ObjectDef<'Root>, ?subsc
         let m = mutation |> function Some (Named n) -> [n] | _ -> []
         let s = subscription |> function Some (Named n) -> [n] | _ -> []
         initialTypes @ s @ m @ schemaConfig.Types |> TypeMap.FromSeq
+           
+    let rec resolveTypeReference (typeDef: TypeDef): TypeDef =
+        match typeDef with
+        | :? ObjectRef as objRef ->
+
+            match typeMap.TryFind<ObjectDef> objRef.Name with
+            | Some realDef ->
+                upcast realDef // TODO: resolveObjectReferences here too
+
+            | None ->
+                failwithf "Referenced type %s not found. Please provide referenced types through SchemaConfig.Types."
+                    objRef.Name
+
+        //        | :? ObjectDef as objDef ->
+        //            upcast resolveObjectReferences (downcast objDef)
+
+        | :? NullableDef as nullDef ->
+            upcast (resolveTypeReference nullDef.OfType).MakeNullable()
+
+        | :? ListOfDef as listOfDef ->
+            upcast (resolveTypeReference listOfDef.OfType).MakeList()
+
+        | _ ->
+            typeDef
+
+    and resolveObjectReferences (objDef: ObjectDef<'a>): ObjectDef<'a> =
+        objDef.WithFields <|
+            Seq.map (fun (KeyValue (_, field)) ->
+                upcast CustomOutputFieldDefinition(field, downcast resolveTypeReference field.TypeDef)
+            ) objDef.Fields
+
+        
+        
+    let query = resolveObjectReferences query
+    let mutation = Option.map resolveObjectReferences mutation
```

Comment:
@johnberzy-bazinga As you can see, the way I went about this is by creating a new `ObjectRef` type that's resolved to real Object Definition as Schema instantiation. I also tinkered with the idea of resolving these references in the Execution Phase, but I thought that could be more error-prone. Let me know if the approach I've taken makes sense to you.

Reply:
Also, this only works when using references at the Root level because it doesn't go into nested Objects. 
Check out lines 186 and 194. I'm getting a casting error in both cases, saying it cannot cast `ObjectDefinition<Planet>` to `ObjectDef<obj>`. 

Any ideas on how to deal with this case? 

I think I'm using the wrong types in one of these functions, but I wanted to ask you before attempting something too hacky.

Reply:
I think it makes sense before the execution step, but a couple of points:

1) The TypeMap.ofSeq call above traverses the tree and I think it will fail unless we [treat the ObjectRef case](https://github.com/fsprojects/FSharp.Data.GraphQL/blob/dev/src/FSharp.Data.GraphQL.Shared/TypeSystem.fs#L1881).
2) I believe We'll need to readd the ObjectDefs to the TypeMap after (or maybe during?) the `resolveObjectReferences` step since we're modifying the fields.

Reply:
@agu-z  Regarding the exception for casting, I think there is no way around that without using reflection. We need to call `resolveObjectReferences<'a>` using reflection where `'a` where is a `Planet` in the specific case.

Alternatively, we can look to add the untyped equivalent of `ObjectDef<'Val>.WithFields` and work with that instead in `resolveObjectReferences`.

Reply:
I think the general approach is right here. If we do it early on in the Schema initialization, we don't need to worry about dealing with ObjectRefs in later phases.

Reply:
Makes sense to me. I apologize for the radio silence. A lot has changed at work and I have been terribly busy. I'll try to take some time next week to finish up this PR.